memory/training/siamese/trainer.py

```diff
@@ -175,3 +168,26 @@ def _setup(self):
         # build train and val data generators
         self._train_gen, self._val_gen = self._build_data_generators(train_dataset, val_dataset)
 
+
+    def neighbor_prediction(self):
+        predictions = []
+
+        cache_dir = self.cache_dir
+        test_img_dir = self.test_dir
+
+        neighbor_gen = NeighborGenerator(self)
+        neighbor_gen.load_data(cache_dir)
+
+        # form [[img_file, [neighbors]], ...]
+        neighbors = neighbor_gen.batch_predict(test_img_dir)
+
+        for pair in neighbors:
+            test_img = np.load(pair[0])['arr_0']
+            for neighbor in pair[1]:
+                neighbor = neighbor[1]
+                n_dir = os.path.join(cache_dir, neighbor)
+                neighbor_img = np.load(n_dir)['arr_0']
+
+                prediction = self._network.predict([np.array([test_img]), np.array([neighbor_img])])
+                predictions.append([pair[0], neighbor, prediction])
```

Comment:
It would be nice if dim 0 of the returned predictions matched up with dim 0 of the input queries. For example, if there are 10 input images, we would look up 10 neighbors for each one and get 10 predictions, so our output shape would be (10, 10). I don't think we need to return the actual query images (in this case coming from `test_img_dir`), because in the more generic predict function format I specified above, the user will anyways be providing those. Also, I would recommend using numpy arrays instead of lists, especially when you have larger arrays, because then you can take advantage of vectorized numpy operations.